vm/src/obj/objiter.rs

```diff
@@ -152,14 +149,31 @@ impl PyValue for PySequenceIterator {
 
 #[pyimpl]
 impl PySequenceIterator {
+    pub fn new_forward(obj: PyObjectRef) -> Self {
+        Self {
+            position: AtomicCell::new(0),
+            obj,
+            reversed: false,
+        }
+    }
+
+    pub fn new_reversed(obj: PyObjectRef, len: isize) -> Self {
+        Self {
+            position: AtomicCell::new(len - 1),
+            obj,
+            reversed: true,
+        }
+    }
+
     #[pymethod(name = "__next__")]
     fn next(&self, vm: &VirtualMachine) -> PyResult {
-        if self.position.get() >= 0 {
+        let pos = self.position.load();
+        if pos >= 0 {
             let step: isize = if self.reversed { -1 } else { 1 };
-            let number = vm.ctx.new_int(self.position.get());
+            let number = vm.ctx.new_int(pos);
             match vm.call_method(&self.obj, "__getitem__", vec![number]) {
                 Ok(val) => {
-                    self.position.set(self.position.get() + step);
+                    self.position.store(pos + step);
```

Comment:
You should not use the store here as this will not be atomic. Another thread can get the old value of `position` and the increment will be wrong. My suggestion is the change the `self.position.load` to `self.position.fetch_add(step)`

Reply:
How about loading it first, and then doing a `compare_exchange` with `prev ± 1`?

Reply:
@palaviv I think the change in the latest commit fixes this, could you take a look?